start breaking up controller manager into two pieces

wlan0 opened the pull request:
This PR addresses: https://github.com/kubernetes/features/issues/88

This commit starts breaking the controller manager into two pieces, namely,
1. cloudprovider dependent piece
2. coudprovider agnostic piece

the controller manager has the following control loops -
- nodeController
- volumeController
- routeController
- serviceController
- replicationController
- endpointController
- resourceQuotaController
- namespaceController
- deploymentController 
  etc..

among the above controller loops,
- nodeController
- volumeController
- routeController
- serviceController

are cloud provider dependent. As kubernetes has evolved tremendously, it has become difficult
for different cloudproviders (currently 8), to make changes and iterate quickly. Moreover, the
cloudproviders are constrained by the kubernetes build/release lifecycle. This commit is the first
step in moving towards a kubernetes code base where cloud providers specific code will move out of
the core repository, and will be maintained by the cloud providers themselves.

I have added a new cloud provider called "external", which signals the controller-manager that
cloud provider specific loops are being run by another controller. I have added these changes in such
a way that the existing cloud providers are not affected. This change is completely backwards compatible, and does not require any changes to the way kubernetes is run today.

Finally, along with the controller-manager, the kubelet also has cloud-provider specific code, and that will be addressed in a different commit/issue.

@alena1108 @ibuildthecloud @thockin @dchen1107 

**Special notes for your reviewer**:

@thockin - Im making this **WIP** PR to ensure that I don't stray too far from everyone's view of how we should make this change. As you can see, only one controller, namely `nodecontroller` can be disabled with the `--cloudprovider=external` flag at the moment. I'm working on cleaning up the `rancher-controller-manger` that I wrote to test this.

Secondly, I'd like to use this PR to address cloudprovider specific code in kubelet and api-server.

**Kubelet**
Kubelet uses provider specific code for node registration and for checking node-status. I thought of two ways to divide the kubelet: 
- We could start a cloud provider specific kubelet on each host as a part of kubernetes, and this cloud-specific-kubelet does node registration and node-status checks. 
- Create a kubelet plugin for each provider, which will be started by kubelet as a long running service. This plugin can be packaged as a binary.

I'm leaning towards the first option. That way, kubelet does not have to manage another process, and we can offload the process management of the cloud-provider-specific-kubelet to something like systemd. 

@dchen1107 @thockin what do you think?

**Kube-apiserver**

Kube-apiserver uses provider specific code for distributing ssh keys to all the nodes of a cluster. Do you have any suggestions about how to address this? 

**Release note**:

``` release-note
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34273)

<!-- Reviewable:end -->

bgrant0607 commented:
@gmarek What do you think about the idea of splitting the node controller into cloud-dependent and cloud-independent controllers?

bgrant0607 commented:
@wlan0 What are the uses in the kubelet? 

Where is the ssh-key-distribution code?

bgrant0607 commented:
ok to test

bgrant0607 commented:
ssh key distribution:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L118
```
			if instances, supported := cloud.Instances(); supported {
				installSSH = instances.AddSSHKeyToAllInstances
			}
```

bgrant0607 commented:
My thought about static information provided by the cloud provider (id, zone, etc.) could be retrieved from a file generated via cloud init or something similar.

I believe cloud-provider-specific volume logic in Kubelet is deprecated, though @saad-ali could confirm or deny that.

saad-ali commented:
> I believe cloud-provider-specific volume logic in Kubelet is deprecated, though @saad-ali could confirm or deny that.

Kubelet by default doesn't do attach/detach (i.e. cloud volume operations), but it retains the ability to do so (enabled via a flag) to support headless operation and thus the dependency still exists.

gmarek commented:
@bgrant0607 - NodeController uses cloud provider read-only and only the most basic operation (listing instances). The sole reason for having cloud provider at all is to be able to quickly delete Nodes that does not exist anymore.

IIUC the reason for this PR is to make it easier to iterate on cloud-provider specific implementations. I think that depending on very basic cloud provider operations (in this case listing instances) should not cause too much trouble for cloud-provider-interface implementers, so keeping the NodeController as it is should be fine. That being said we can create a separate controller that does the removal and remove this functionality from NodeController, thus making it cloud-agnostic, if it proves not to be the case.

wlan0 commented:
@bgrant0607 - As @saad-ali mentioned, there is a cloud provider dependency for kubelet. 

@gmarek Thanks for the response. I'll leave nodecontroller as it is. 

@bgrant0607 I'm pushing to propose this as a feature to kubernetes-1.5, and basing the proposal off of this issue - https://github.com/kubernetes/features/issues/88 

Should I make you the assignee for the feature proposal?

Commit:
add cloud-controller-manager as the first step in breaking controller-manager

wlan0 commented:
@bgrant0607 @thockin - Here's a reference to the external controller manager that i'm using for the rancher cloud provider - https://github.com/wlan0/kubernetes/tree/to_show/cmd/external-controller-manager 

The cloud provider is linked here (standard for all cloud providers) - https://github.com/wlan0/kubernetes/blob/to_show/cmd/external-controller-manager/controller-manager.go#L35

The cloud provider needs to satisfy this interface - https://github.com/wlan0/kubernetes/blob/to_show/cmd/external-controller-manager/rancher/rancher.go#L766

Everything else is a simple copy+paste of existing code.

wlan0 commented:
@bgrant0607 - I managed to successfully run the service, and route controller loops as well, along with the node controller loop. I also tested these and they work just fine. 

It's here - https://github.com/wlan0/kubernetes/tree/to_show/cmd/external-controller-manager

I'll be trying to run the attach_detach controller loop in the cloud specific controller-manager now, and as @thockin suggested, this is likely where I'll encounter challenges. 

Also, the deadline for kubernetes-1.5 feature proposal has passed, so we don't have to worry about pushing for a feature for now.

thockin commented:
I do think we should aim for a 100% cloud-neutral controller-manager.

I told you kubelet would be the hardest, didn't I?  :)  Breaking kubelet into two processes is going to run counter to all the simplified UX work.  We'll need to think really hard about that.  @mikedanese 

Volume attach/detach is part of the volumes API, so it should already be a little insulated.  I would leave this for near last.

I guess what I want to see most is an exploration of all the controller-manager controllers to see what other things we have not thought of that will be extra hard...

wlan0 commented:
> I told you kubelet would be the hardest, didn't I? :) Breaking kubelet into two
> processes is going to run counter to all the simplified UX work. We'll need to
> think really hard about that. @mikedanese

@thockin @mikedanese Is the UX simplification work mentioned above https://github.com/kubernetes/features/issues/11?

> I guess what I want to see most is an exploration of all the controller-manager
> controllers to see what other things we have not thought of that will be extra hard...

Do you mean concerns arising from running these controllers as separate services?

For eg, when nodecontroller and servicecontroller were on different controller-managers, pod creation took longer. The problem disappeared when both were on the same controller-manager. 

I could investigate these and other controllers to identify locking issues, potential deadlocks, leader election, performance issues etc.

thockin commented:
yes, exactly.  But not just what happens when they run separately, but
actually walk through the main controller loops and think about anything
that will be extra hard to disentangle.  I am SURE something is lurking
that will make you say "oh damn, that's tricky" - I'd rather try to find it
sooner.

Or maybe the best way forward is just to try.  Shoot for a really hacky
prototype level of quality, and see how hard it is.  Maybe this really will
be just a refactoring problem, and not need and architectural decisions.  I
doubt it, but maybe... :)

On Thu, Oct 13, 2016 at 3:45 PM, Sidhartha Mani <notifications@github.com>
wrote:

> I told you kubelet would be the hardest, didn't I? :) Breaking kubelet
> into two
> processes is going to run counter to all the simplified UX work. We'll
> need to
> think really hard about that. @mikedanese <https://github.com/mikedanese>
>
> @thockin <https://github.com/thockin> @mikedanese
> <https://github.com/mikedanese> Is the UX simplification work mentioned
> above kubernetes/features#11
> <https://github.com/kubernetes/features/issues/11>?
>
> I guess what I want to see most is an exploration of all the
> controller-manager
> controllers to see what other things we have not thought of that will be
> extra hard...
>
> Do you mean concerns arising from running these controllers as separate
> services?
>
> For eg, when nodecontroller and servicecontroller were on different
> controller-managers, pod creation took longer. The problem disappeared when
> both were on the same controller-manager.
>
> I could investigate these and other controllers to identify locking
> issues, potential deadlocks, leader election, performance issues etc.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34273#issuecomment-253660390>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVLBVeo_MloG1T6qQHgWXOmdGQgmgks5qzrSOgaJpZM4KQa1j>
> .
>

gmarek commented:
@thockin - do you think we should split out removing Nodes not visible in cloud provider out of NC to make it cloud agnostic?

thockin commented:
yes, I'd like to totally kill off cloud provider, completely.

gmarek commented:
OK, SG @thockin - please file an issue and prioritize accordingly.

thockin commented:
One step at a time.  Let's let this PR make some progress before we all
jump in :)

On Wed, Oct 19, 2016 at 12:37 AM, Marek Grabowski <notifications@github.com>
wrote:

> OK, SG @thockin <https://github.com/thockin> - please file an issue and
> prioritize accordingly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34273#issuecomment-254735810>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVH-08CXxQJJFQxPoOZb7EW0ju6Mfks5q1cjNgaJpZM4KQa1j>
> .
>

wlan0 commented:
@thockin - I have a prototype level quality of code where I disentangled cloud provider specific controllers from the rest (except volumes). 

In my design, I made the decision to run entire controller loops in the provider specific binary, rather than only the very minute parts that rely on cloud provider. If I drilled down further into the controller loops, it would be significantly harder to disentangle, due to very tight coupling of the cloud provider specific code with the rest of the data structures/functions within a controller.

This approach makes the implementation really simple. There are four controllers that rely on cloud provider specific code. These are node controller, service controller, route controller and attach detach controller (volumes, which i have left for the very last). I created copies of each of these controllers, and bundled them together into one binary. This binary registers itself as a controller, and runs the cloud specific controller loops with the user-agent named "external-controller-manager". 

After analyzing the cloud specific controllers, this is what I found - 

1. All of the Informers get events from the REST API, similarly all writes to resources are done directly using the REST API. There is no local state maintained, therefore there is no issue with running another set of controller loops as another binary.

2. Leader election ID for the new binary - This new binary will require a new leader election endpoint name. I just called it "external-controller-manager" in my prototype. As long as this is done, it should have no issues setting up HA external-controller-manager.

3. Event broadcast/Sink framework - This framework also talks to the REST API. The shared instance of this is only used by the attach_detach_controller to send events to the api-server. The other controllers create their own instances of the broadcaster. I can't spot any risk of concurrency/disentanglement issues with this part of the system. 

4. It increases the number of queries to check the state of resources in the server. In the previous model, all of the controllers shared the informer, and now there will be another informer for the new binary making requests to the server. The impact on performance is not noticeable for small deployments. I haven't tested beyond that.

To me, the self-contained, modular nature of the various subsystems makes it look straight forward. But, If there is something missing, please do let me know. I would love to investigate further. 

I haven't analyzed the rest of the controllers. I'll probably uncover more issues once I start breaking up kubelet and the volume controllers. Will sync back here once I do that.

wlan0 commented:
@thockin I was worried earlier that it was too simple. Well, not anymore! (*I'm counter intuitively happy*  \_(ツ)_/¯)

The kubelet is tricky. It is tightly coupled with the cloud provider. It uses the cloud provider to

1. Figure out the cloud nodename of the host it is running on
    * First to check for existing kubelet config for that nodename (https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L165)
    * Then to instantiate a new Kubelet object (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L286)

2. Performs various cloud provider calls while creating the initial Node object (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L218)

3. Periodically poll the cloud provider to figure out if the node has any new IP addresses associated with it (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L372)

4. It allows provider to perform network configuration (only in case of GCE) (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_network.go#L88)

5. It allows provider to post process DNS settings (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_network.go#L142)

Now, the reason for the trickiness..

 * **Approach 1** The first approach I considered was to create two kubelets just like I did for controller-manger. Unlike the controller-manger, running two kubelets (one generic and one cloud provider specific) is not easy. Node object is initially created by the kubelet, so this would have to be created before any of the kubelet loops are started. Since creation is cloud provider dependent, we would have to co-ordinate the two kubelets such that the node object is created first and then the various loops are run on it. Moreover, this approach will complicate the node creation method and make it less readable. 

 *  **Approach 2** The Second approach I considered was to create a cloudprovider plugin (like a CNI plugin) as a binary. Then, each call to the cloud provider would actually execute a binary. Vendors can package their own binary, as along as it adheres to a certain interface. The binary would have to initialize the cloudprovider object on each execution by reading a config file (currently this is how it is initialized, but its done only once on startup of kubelet). This poses performance concerns, since there is a periodic update loop for updating Node addresses.

 *  **Approach 3** The third approach would be to link the cloud provider to the kubelet, and have a different kubelet for each cloud provider. This would require me to create a library out of the kubelet operations. Various vendors would use that library and create a binary that "imports" their own cloudprovider pkg and the kubelet library pkg. This requires that only one process be run (unlike the above two approaches). 

I find **Approach 3** to be the best, and I've started working on it. Let me know if you agree with this. I'd love to hear better approaches, or any other feedback on this.

wlan0 commented:
@thockin any thoughts?

thockin commented:
I think your model sounds like an OK start.

Routes and service seem to be entirely cloud-specific.  Node seems to only
use cloud provider for one very specific thing, so probably it should be
split into two parts - one generic and one not.

Some thoughts from grepping around:

plugin/pkg/admission/persistentvolume/label/admission.go uses
cloudprovider.  This one will be tricky.  What will happen if this stops
being an admission controller and becomes async instead?  @justinsb

pkg/genericapiserver/config.go is weird.  I can't immediately explain it.

Ignore pkg/volume for now.  In fact, ignore any case that uses a specific
cloud provider - focus on killing the cloudprovider.Interface



On Thu, Oct 20, 2016 at 6:16 PM, Sidhartha Mani <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> - I have a prototype level quality
> of code where I disentangled cloud provider specific controllers from the
> rest (except volumes).
>
> In my design, I made the decision to run entire controller loops in the
> provider specific binary, rather than only the very minute parts that rely
> on cloud provider. If I drilled down further into the controller loops, it
> would be significantly harder to disentangle, due to very tight coupling of
> the cloud provider specific code with the rest of the data
> structures/functions within a controller.
>
> This approach makes the implementation really simple. There are four
> controllers that rely on cloud provider specific code. These are node
> controller, service controller, route controller and attach detach
> controller (volumes, which i have left for the very last). I created copies
> of each of these controllers, and bundled them together into one binary.
> This binary registers itself as a controller, and runs the cloud specific
> controller loops with the user-agent named "external-controller-manager".
>
> After analyzing the cloud specific controllers, this is what I found -
>
>    1.
>
>    All of the Informers get events from the REST API, similarly all
>    writes to resources are done directly using the REST API. There is no local
>    state maintained, therefore there is no issue with running another set of
>    controller loops as another binary.
>    2.
>
>    Leader election ID for the new binary - This new binary will require a
>    new leader election endpoint name. I just called it
>    "external-controller-manager" in my prototype. As long as this is done, it
>    should have no issues setting up HA external-controller-manager.
>    3.
>
>    Event broadcast/Sink framework - This framework also talks to the REST
>    API. The shared instance of this is only used by the
>    attach_detach_controller to send events to the api-server. The other
>    controllers create their own instances of the broadcaster. I can't spot any
>    risk of concurrency/disentanglement issues with this part of the system.
>    4.
>
>    It increases the number of queries to check the state of resources in
>    the server. In the previous model, all of the controllers shared the
>    informer, and now there will be another informer for the new binary making
>    requests to the server. The impact on performance is not noticeable for
>    small deployments. I haven't tested beyond that.
>
> To me, the self-contained, modular nature of the various subsystems makes
> it look straight forward. But, If there is something missing, please do let
> me know. I would love to investigate further.
>
> I haven't analyzed the rest of the controllers. I'll probably uncover more
> issues once I start breaking up kubelet and the volume controllers. Will
> sync back here once I do that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34273#issuecomment-255269620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVP3Ucs7ywTB2Iqwm5vDVCX3aWAegks5q2BJjgaJpZM4KQa1j>
> .
>

thockin commented:
It seems that the volume controllers only need cloudprovider to satisfy
volume interface, so we will need to tackle pkg/volume.  Volumes need cloud
providers, but they only need SPECIFIC cloud providers.  The real problem
is that cloud providers can take params from (e.g. cloud config).  That's a
tricky one...

On Thu, Oct 20, 2016 at 9:00 PM, Tim Hockin <thockin@google.com> wrote:

> I think your model sounds like an OK start.
>
> Routes and service seem to be entirely cloud-specific.  Node seems to only
> use cloud provider for one very specific thing, so probably it should be
> split into two parts - one generic and one not.
>
> Some thoughts from grepping around:
>
> plugin/pkg/admission/persistentvolume/label/admission.go uses
> cloudprovider.  This one will be tricky.  What will happen if this stops
> being an admission controller and becomes async instead?  @justinsb
>
> pkg/genericapiserver/config.go is weird.  I can't immediately explain it.
>
> Ignore pkg/volume for now.  In fact, ignore any case that uses a specific
> cloud provider - focus on killing the cloudprovider.Interface
>
>
>
> On Thu, Oct 20, 2016 at 6:16 PM, Sidhartha Mani <notifications@github.com>
> wrote:
>
>> @thockin <https://github.com/thockin> - I have a prototype level quality
>> of code where I disentangled cloud provider specific controllers from the
>> rest (except volumes).
>>
>> In my design, I made the decision to run entire controller loops in the
>> provider specific binary, rather than only the very minute parts that rely
>> on cloud provider. If I drilled down further into the controller loops, it
>> would be significantly harder to disentangle, due to very tight coupling of
>> the cloud provider specific code with the rest of the data
>> structures/functions within a controller.
>>
>> This approach makes the implementation really simple. There are four
>> controllers that rely on cloud provider specific code. These are node
>> controller, service controller, route controller and attach detach
>> controller (volumes, which i have left for the very last). I created copies
>> of each of these controllers, and bundled them together into one binary.
>> This binary registers itself as a controller, and runs the cloud specific
>> controller loops with the user-agent named "external-controller-manager".
>>
>> After analyzing the cloud specific controllers, this is what I found -
>>
>>    1.
>>
>>    All of the Informers get events from the REST API, similarly all
>>    writes to resources are done directly using the REST API. There is no local
>>    state maintained, therefore there is no issue with running another set of
>>    controller loops as another binary.
>>    2.
>>
>>    Leader election ID for the new binary - This new binary will require
>>    a new leader election endpoint name. I just called it
>>    "external-controller-manager" in my prototype. As long as this is done, it
>>    should have no issues setting up HA external-controller-manager.
>>    3.
>>
>>    Event broadcast/Sink framework - This framework also talks to the
>>    REST API. The shared instance of this is only used by the
>>    attach_detach_controller to send events to the api-server. The other
>>    controllers create their own instances of the broadcaster. I can't spot any
>>    risk of concurrency/disentanglement issues with this part of the system.
>>    4.
>>
>>    It increases the number of queries to check the state of resources in
>>    the server. In the previous model, all of the controllers shared the
>>    informer, and now there will be another informer for the new binary making
>>    requests to the server. The impact on performance is not noticeable for
>>    small deployments. I haven't tested beyond that.
>>
>> To me, the self-contained, modular nature of the various subsystems makes
>> it look straight forward. But, If there is something missing, please do let
>> me know. I would love to investigate further.
>>
>> I haven't analyzed the rest of the controllers. I'll probably uncover
>> more issues once I start breaking up kubelet and the volume controllers.
>> Will sync back here once I do that.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/34273#issuecomment-255269620>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AFVgVP3Ucs7ywTB2Iqwm5vDVCX3aWAegks5q2BJjgaJpZM4KQa1j>
>> .
>>
>
>

wlan0 commented:
@thockin - some updates...

> pkg/genericapiserver/config.go is weird. I can't immediately explain it.

I looked into this. This fills in the value of `--external-hostname` for the apiserver if the user doesn't set it. This is used by cluster federation for cluster registration (Cluster gets registered by externalHostname). However, there are other ways for the user to provide the ExternalIP, such as using a DNS name, or configure a static value. Therefore, I think we should remove cloudprovider reference from here. 

In the rest of the API server, there is only one reference to cloudprovider - 

https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L110

This code is used to setup SSH tunneling for master -> node communications (for GCE only). The SSH tunnel secures the otherwise insecure connection from apiserver to kubelet/node. There is a new feature now to secure this communication - https://github.com/kubernetes/kubernetes/pull/34381. Therefore, this code can also be safely removed. 

kube-apiserver only needs the above two fixes to become cloudprovider agnostic.

Here's my simple roadmap for completing this 

 - [x] Cloud Specific Controller Manager (without killing cloudprovider.Interface)   
 - [x] Cloud Specific kubelet (without killing cloudprovider.Interface)
 - [x] Cloud agnostic kube-apiserver (without killing cloudprovider.Interface)
 - [ ] Volumes
 - [ ] Kill cloudprovider.Interface

justinsb commented:
> plugin/pkg/admission/persistentvolume/label/admission.go uses cloudprovider.  This one will be tricky.  What will happen if this stops being an admission controller and becomes async instead?  @justinsb

Race conditions, probably.  It's actually come up as a problem before, because it puts a synchronous cloud-API call into the k8s API object-creation path, which means that if you're rate-limited you can bust our perf targets.

We need the zone information so that we can schedule pods onto nodes in the same zone.  If the volume comes up without being labeled, I think we could schedule a pod into a zone that can't mount the volume.  And then we would have no way to recover from that - the pod would just sit there failing to start.

We _could_ move this into the scheduler, but that feels worse.

Likely if you want to decouple this, the correct solution is to somehow mark the volume as "not ready" in the admission controller, and then have the async controller add the annotations and mark the volume as ready.  Scheduling would be blocked until a volume was ready.  I don't know if we have a notion of volume readiness though? 

Another option might be to do this at claim-binding time, which feels less right, but is probably simpler.

> pkg/genericapiserver/config.go is weird.  I can't immediately explain it.

It looks to be setting the addresses & hostname correctly for the cloudprovider.  I can't explain why this particular code block is here (just to set the defaults for the flags?) but we do this elsewhere as part of initial node registration.   As of 1.4 we had the requirement that the hostname be resolvable, for example, so we had to set the NodeName to be a particular value so that everything would both work on the master and be resolvable through the API.  This should finally be fixed in 1.5 though with #33718 .  Similarly we need to discover the IP addresses and cloud labels to apply to the node.

I guess if you wanted to move this out of the kubelet, it could be an admission controller!  And we do have a notion of node readiness, so we could have a "processed by cloud provider" readiness state.

----

It does feel like we could solve a lot of problems if we had external admission controllers.

wlan0 commented:
> It looks to be setting the addresses & hostname correctly for the cloudprovider. I can't explain why this particular code block is here (just to set the defaults for the flags?) but we do this elsewhere as      part of initial node registration. 

This is done to make the apiserver accessible externally. AFAIK, This information is *not* used to fill in the node object. Node registration is done by the kubelet.

> I guess if you wanted to move this out of the kubelet, it could be an admission controller! And we do have a notion of node readiness, so we could have a "processed by cloud provider" readiness state.

This code is called from the apiserver. Not the kubelet.

wlan0 commented:
> We could move this into the scheduler, but that feels worse.

Agreed.

> Likely if you want to decouple this, the correct solution is to somehow mark the volume as "not ready" in the admission controller, and then have the async controller add the annotations and mark the volume as ready. Scheduling would be blocked until a volume was ready. I don't know if we have a notion of volume readiness though?

@justinsb A volume is ready when it is "Available". If we set the phase of the volume as "Pending" until an external admission controller sets the labels on the volume object, and then update it to "Available", this would work. However, if we needed to add another external admission controller for the same object, then there would be race condition (Who will update it to "Available"?) 

The other approach would be to leave the volume phases as they are and change the scheduler to look for an annotation (for eg. "kubernetes.io.cloud.labels.needed": "false"). If this annotation is present, schedule the volumes. An external admission controller will asynchronously set the cloudprovider labels and update this annotation. In this approach, we can have multiple external admission controllers, as long as there are define different annotation for different controllers. 

Both of these approaches block scheduling until the volumes are marked as ready.

wlan0 commented:
@thockin I removed the WIP. This PR is ready to be reviewed. 

I created a new controller that does the cloudprovider specific parts of the existing node controller. This is the only change needed run a cloud specific controller manager.

Note: this still doesn't address volumes.

luxas commented on pkg/controller/cloud/cloud_node_controller.go:
```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
nit: 2016

wlan0 replied:
Fixed.

luxas commented on pkg/controller/cloud/cloud_node_controller.go:
```diff
@@ -0,0 +1,149 @@
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
```
just a heads-up; soon we'll use client-go

luxas commented on pkg/controller/cloud/cloud_node_controller.go:
```diff
@@ -0,0 +1,149 @@
+				node := &nodes.Items[i]
+				for rep := 0; rep < nodeStatusUpdateRetry; rep++ {
+					_, currentReadyCondition = api.GetNodeCondition(&node.Status, api.NodeReady)
+					if err == nil {
```
Why are you checking for `err == nil` here?
The last time err got a value is on line `node, err = cnc.kubeClient.Core().Nodes().Get(name)`, why don't handle that case there or am I missing something?

wlan0 replied:
Fixed.

luxas reviewed (commented):
So the purpose of this is to split out the cloud-specific parts from `pkg/controller/node/nodecontroller.go` right and use this "interface" in controller-manager instead, right?

wlan0 commented:
> So the purpose of this is to split out the cloud-specific parts from   pkg/controller/node/nodecontroller.go right and use this "interface" in controller-manager instead, right?

@luxas that is absolutely correct. 

I've updated the PR with the changes.

luxas commented:
@wlan0 Can you write a proposal for what you're planning to do.
It will be easier to spread the word about this that way...

wlan0 commented:
@luxas yes, here's the proposal - https://github.com/kubernetes/kubernetes/pull/37037

thockin commented:
This code doesn't seem to be called from anywhere?

wlan0 commented:
> This code doesn't seem to be called from anywhere?

This is just a library for now. We should create a new repository under kubernetes org where the cloud-controller-manager will exist.

thockin commented:
I am wary of checking in libs that literally nobody uses yet.  How bad would it be to keep this as a commit and build this PR towards a fully-formed split?  Step 0 would be just to run 2 controllers that are identical in behavior to what we have today, and work out the issues there.   Then we can kick the cloud providers out..

luxas commented:
@wlan0 Why couldn't you make the current controller-manager use `CloudNodeController` right now directly? Doesn't  `CloudNodeController` support all the features controller-manager has built-in right now?

Otherwise we could just use `CloudNodeController` directly from controller-manager today in order to reduce the complexity and at the same time add the `external` cloud provider that doesn't run that code.

I guess I'm missing something here (haven't read the functionality of this code vs the current implementation line-by-line, but I guess it's pretty the same).

WDYT?

wlan0 commented:
@thockin Do you mean that I should create another controller-manager called fake-cloud-controller-manager, which takes in `--cloudprovider=${CLOUD_PROVIDER}` as an option for the time being? Then kick the cloudprovider out of this cloud-controller-manager once we iron out all the issues?

Note: I called it fake-cloud-controller-manager because it'll get merged in with an expectation that its not meant for the long term. If I called it cloud-controller-manager, then I feel like we will need to remove it/rename it later. 

@luxas I feel like adding `CloudNodeController` to the existing controller-manager seems like a good idea. However, running this loop as a part of the same binary still won't be the same as our final goal. If this runs in the same binary, then it'll use the same shared informer, and we might not see potential timing issues.

Commit:
start breaking up controller manager into two pieces

Addresses: kubernetes/features#88

This commit starts breaking the controller manager into two pieces, namely,

1. cloudprovider dependent piece
2. coudprovider agnostic piece

the controller manager has the following control loops -

   - nodeController
   - volumeController
   - routeController
   - serviceController
   - replicationController
   - endpointController
   - resourcequotacontroller
   - namespacecontroller
   - deploymentController etc..

among the above controller loops,

   - nodeController
   - volumeController
   - routeController
   - serviceController

are cloud provider dependent. As kubernetes has evolved tremendously, it has become difficult
for different cloudproviders (currently 8), to make changes and iterate quickly. Moreover, the
cloudproviders are constrained by the kubernetes build/release lifecycle. This commit is the first
step in moving towards a kubernetes code base where cloud providers specific code will move out of
the core repository, and will be maintained by the cloud providers themselves.

Finally, along with the controller-manager, the kubelet also has cloud-provider specific code, and that will
be addressed in a different commit/issue.

wojtek-t commented:
@gmarek

luxas commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,253 @@
+// nodes.
+//
+// CAUTION: If you update code in this file, you may need to also update code
+//          in contrib/mesos/pkg/controllermanager/controllermanager.go
```
I don't think this exists more

luxas commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,253 @@
+)
+
+// NewControllerManagerCommand creates a *cobra.Command object with default parameters
+func NewControllerManagerCommand() *cobra.Command {
```
Can we be consistent with referring to this application as `CloudControllerManager` to avoid confusion with the "normal" controller-manager?

luxas commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,253 @@
+	}
+	sharedInformers := informers.NewSharedInformerFactory(client("shared-informers"), nil, ResyncPeriod(s)())
+
+	_, clusterCIDR, err := net.ParseCIDR(s.ClusterCIDR)
```
Is there any way we can try to fetch this value from the api-server instead of having it as a flag

I really would want to keep the flags down here

thockin replied:
API server doesn't have cluster CIDR yet...

luxas commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -0,0 +1,95 @@
+
+// Package options provides the flags used for the controller manager.
+//
+// CAUTION: If you update code in this file, you may need to also update code
```
Same, you can just remove these

luxas commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -0,0 +1,95 @@
+limitations under the License.
+*/
+
+// Package options provides the flags used for the controller manager.
```
cloud controller manager

luxas commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -0,0 +1,95 @@
+			ControllerStartInterval: metav1.Duration{Duration: 0 * time.Second},
+			EnableGarbageCollector:  true,
+			ConcurrentGCSyncs:       20,
+			ClusterSigningCertFile:  "/etc/kubernetes/ca/ca.pem",
```
These files aren't populated by default in the cluster, right?
Why not just use the service account by default in order to connect to the apiserver?

thockin replied:
I don't know what the Signing cert is for, but we should strip out everything we don't need.

luxas commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -0,0 +1,95 @@
+func NewExternalCMServer() *ExternalCMServer {
+	s := ExternalCMServer{
+		KubeControllerManagerConfiguration: componentconfig.KubeControllerManagerConfiguration{
+			Port:                    ports.ControllerManagerPort,
```
Is it worthwhile to increment the port for the cloud-controller-manager so it's easy to distinguish?

thockin replied:
yeah, you should allocate your own.

luxas commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -0,0 +1,95 @@
+func (s *ExternalCMServer) AddFlags(fs *pflag.FlagSet) {
+	fs.Int32Var(&s.Port, "port", s.Port, "The port that the controller-manager's http service runs on")
+	fs.Var(componentconfig.IPVar{Val: &s.Address}, "address", "The IP address to serve on (set to 0.0.0.0 for all interfaces)")
+	fs.StringVar(&s.CloudProvider, "cloud-provider", s.CloudProvider, "The provider of cloud services. Empty for no provider.")
```
Hmm, why is this here?
Will it be possible to build in many cloudproviders in the same cloudcontrollermanager?
For example, I could make a cloudcontrollermanager myself that would include both `raspberrypi-cloud` and `rancher` as cloudproviders and I could just toggle between those in my cluster?

Or do we want to make it clear that only one cloudprovider is supported per cloudcontrollermanager?

thockin replied:
step 1 - move the existing code
step 2 - fix it

We're still barely scratching step 1. :)

luxas replied:
@thockin Yes, fully agree, just giving some food for thought for future improvements, follow-ups :smile:

luxas commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -378,7 +378,9 @@ func StartControllers(controllers map[string]InitFunc, s *options.CMServer, root
 
 	if s.AllocateNodeCIDRs && s.ConfigureCloudRoutes {
-		if cloud == nil {
+		if s.CloudProvider == "external" {
```
Can you make `external` a constant somewhere?

luxas replied:
ping, seems like you missed this :)

wlan0 replied:
Thanks, looks like kube clientset library was renamed as well.

luxas reviewed (commented):
Looks good to me at a first glance, but I'm still wondering why we're including the existing 7 cloudproviders into the `cmd/cloud-controller-manager` binary.

Will the "official" cloud-controller-manager binary serve as a way to transition to the new flow while still in core until everything's splitted out?

luxas reviewed (approved):
Ok, so I talked with @wlan0, and we agreed that the best way forward here is to merge this PR, and address comments in follow-ups.

Otherwise @wlan0 is stalled for longer time than what's preferable. Also, others can start chiming in on the work when this MVP is merged.

@thockin @dchen1107 @bgrant0607 Ok to merge this and that way be able to proceed on this effort?

alena1108 commented:
Agree with @luxas. All providers that are currently not a part of k8s repo (Rancher, DigitalOcean as an example), will get unblocked as soon as this PR gets in. Once this PR is merged, Rancher will start using libraries created by @wlan0. It means ongoing testing of the path + contributing further fixes. 

@wlan0 I see some failing build, is it something we should resolve?

joonas commented:
Hey, wanted to chime in our support here as well since we at DigitalOcean are also very interested in seeing this work moving along since it has become apparent that this would be the way forward for future cloud providers.

luxas commented:
Cool, thanks for chiming in @alena1108 and @joonas 

@alena1108 No, it's just the CRI test that's flaking, not related to this PR. It has been down for quite a long time now, but it's not blocking so don't worry.

thockin commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,253 @@
+	s.AddFlags(pflag.CommandLine)
+	cmd := &cobra.Command{
+		Use: "cloud-controller-manager",
+		Long: `The Kubernetes controller manager is a daemon that embeds
```
This description is a little chatty and the examples don't apply to cloud..

thockin commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,253 @@
+	}
+	leaderElectionClient := clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "leader-election"))
+
+	go func() {
```
The total lack of comments in this file bums me out...

wlan0 replied:
Sorry, I took care of it now.

thockin commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,253 @@
+	return cmd
+}
+
+func ResyncPeriod(s *options.ExternalCMServer) func() time.Duration {
```
comment

thockin commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,253 @@
+	"k8s.io/kubernetes/cmd/cloud-controller-manager/app/options"
+	"k8s.io/kubernetes/pkg/api/v1"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5"
+	v1core "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5/typed/core/v1"
```
@caesarxuchao should this be using client-go?

caesarxuchao replied:
No. For now we use release_1_5 in the code base.

thockin commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -0,0 +1,95 @@
+	"github.com/spf13/pflag"
+)
+
+// ExternalCMServer is the main context object for the controller manager.
```
Can we sanitize these names?  Why "External" ?  "CM" is both insufficient and wrong.

thockin commented on pkg/cloudprovider/plugins.go:
```diff
@@ -90,6 +90,11 @@ func InitCloudProvider(name string, configFilePath string) (Interface, error) {
 	var cloud Interface
 	var err error
 
+	if name == "external" {
```
I wonder if this distinction is just noise - maybe we should just take "" as "external" ?

wlan0 replied:
Yes, this is just noise and can be removed. I'll update the proposal as well

thockin commented on pkg/controller/cloud/nodecontroller.go:
```diff
@@ -0,0 +1,154 @@
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+type CloudNodeController struct {
```
rather than forking this, which means we'll have to constantly sync between them, maybe we can use the same code in both controllers?

wlan0 replied:
@thockin this would be a good idea in general. In this case however the logic on the CloudNodeController side is so minimal and in order to share code, I'd have to break down the existing NodeController logic into smaller pieces and glue them together. I'd rather not touch NodeController.

thockin reviewed (commented):
I like Lucas's comments, too.

luxas commented:
@thockin Is that a LGTM?
I know @wlan0 is committed to moving this forward, and so am I and others.

Your comments @thockin as well as mine were more nits than huge structural changes, so I propose to continue by merging, unless the other reviewers has something against it.

thockin commented:
It's not LGTM - there are a bunch of comments that need to be resolved.

On Tue, Dec 13, 2016 at 11:45 AM, Lucas Käldström <notifications@github.com>
wrote:

> *@luxas* commented on this pull request.
> ------------------------------
>
> In cmd/cloud-controller-manager/app/options/options.go
> <https://github.com/kubernetes/kubernetes/pull/34273>:
>
> > +			ControllerStartInterval: metav1.Duration{Duration: 0 * time.Second},
> +			EnableGarbageCollector:  true,
> +			ConcurrentGCSyncs:       20,
> +			ClusterSigningCertFile:  "/etc/kubernetes/ca/ca.pem",
> +			ClusterSigningKeyFile:   "/etc/kubernetes/ca/ca.key",
> +		},
> +	}
> +	s.LeaderElection.LeaderElect = true
> +	return &s
> +}
> +
> +// AddFlags adds flags for a specific ExternalCMServer to the specified FlagSet
> +func (s *ExternalCMServer) AddFlags(fs *pflag.FlagSet) {
> +	fs.Int32Var(&s.Port, "port", s.Port, "The port that the controller-manager's http service runs on")
> +	fs.Var(componentconfig.IPVar{Val: &s.Address}, "address", "The IP address to serve on (set to 0.0.0.0 for all interfaces)")
> +	fs.StringVar(&s.CloudProvider, "cloud-provider", s.CloudProvider, "The provider of cloud services. Empty for no provider.")
>
> @thockin <https://github.com/thockin> Yes, fully agree, just giving some
> food for thought for future improvements, follow-ups 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34273>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVDFLheeNG8C20YXYzSenPHt3N_bvks5rHvXvgaJpZM4KQa1j>
> .
>

luxas commented:
Ok, @wlan0 please fix the nits/comments described here then
On a high level, fix:
 - Comments (add many and remove the ones that are wrong)
 - Interpret `""` as `external`
 - The long description
 - Allocate a new port

You can do those changes in a third commit so it's easy to see the changes

Commit:
sanitize names and add more comments, and other essential boilerplate changes

wlan0 commented:
@thockin  @luxas this is ready to be reviewed

luxas reviewed (approved):
LGTM

davidopp commented:
Note: We're going to be making some nontrivial changes to NodeController once @gmarek picks up and finishes #34825, which should be mid-January. Could you please try to avoid any major changes to NodeController until his changes get merged?

luxas commented:
@davidopp Sure! We won't touch the normal controller-manager this release @wlan0, will we?

@davidopp Are there any cloud-specific things to take into account in #34825?
I guess we have to update the forked cloud variant for NodeController later then if the logic changes a lot, but I don't think this will affect controller-manager yet.

Thanks for the heads-up though!

wlan0 commented:
@davidopp @luxas - In this and the next release, we have planned to leave the controller manager as it is. In the releases after that, we'll deprecate the `--cloud-provider` flag and subsequently remove it from the controller manager.

liggitt commented:
cc @deads2k

thockin commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,251 @@
+}
+
+// ResyncPeriod computes the time interval a shared informer waits before resyncing with the api server
+func ResyncPeriod(s *options.CloudControllerManagerServer) func() time.Duration {
```
could be private ?

thockin reviewed (approved):
LGTM

luxas commented:
Yay!
Thanks @thockin for that christmas :gift: :)

liggitt commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -0,0 +1,251 @@
+			ClientConfig: kubeconfig,
+		}
+		var clientBuilder controller.ControllerClientBuilder
+		if len(s.ServiceAccountKeyFile) > 0 && s.UseServiceAccountCredentials {
```
This controller doesn't need ServiceAccountKeyFile, and should probably just use the UseServiceAccountCredentials flag to determine this... the goal here is to run each controller loop with a distinct identity so its permissions can be scoped to its job.

thockin commented:
AI to @wlan0: I tagged this as needing a relnote.  Anyone who does a cluster setup needs to know this is coming.

@justinsb @luxas FYI in case we want to optionally enable this (certainly not by default, and maybe not at all).

We need a doc that explains "This exists, it is pre-alpha, but here's how you can manually test it"

thockin commented:
After discussion, ignore the relnote.  We do not want mere mortals using this yet.

luxas commented:
@thockin PTAL at https://docs.google.com/document/d/1bpEGhdg8t-2Ae-IL_ezZIPklMHpZLWwnoAPMgLwT3oU

wlan0 commented:
The bazel build failure is a flake.

`./hack/verify-bazel.sh` script succeeds.
`./hack/update-bazel.sh` does not modify any files.

There are other PRs at the moment which are failing due to the bazel failure. Last successful bazel build was 5 hrs ago.

thockin commented:
@mikedanese @ixdy re bazel failures.

@wlan0 no way am I hand-merging this one :)

wlan0 commented:
@k8s-bot bazel test this

luxas commented:
@k8s-bot bazel test this
